File: [0046][Medium][Permutations][backtracking]/Permutations_2.py
# leetcode time     cost : 36 ms
# leetcode memory   cost : 13.8 MB
# Time  Complexity: O(n*n!)
# Space Complexity: O(n)
# solution 1,backtracking
from typing import List
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def trackBack(nums,track):
            if len(track) == len(nums):
                res.append(track[:]) # 需要传递下track的拷贝，否则对track的修改会影响到结果
                return
            for i in nums:
                if i in track:
                    continue
                track.append(i)
                trackBack(nums,track)
                track.pop()
        res = []
        track = []
        trackBack(nums,track)
        return res
def main():
    inputX,expectRes = [1,2,3], [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,2,1],[3,1,2]]
    obj = Solution()
    result = obj.permute(inputX)
    try:
        assert result == expectRes
        print("passed, result is follow expect:",result)
    except AssertionError as aError:
        print('failed, result >> ', result,"<< is wrong, ","expect is : "+ expectRes)
    
if __name__ =='__main__':
    main()  

File: [0046][Medium][Permutations][backtracking]/Permutations_2.cpp
// -*- coding: utf-8 -*- 
// leetcode time     cost : 4 ms
// leetcode memory   cost : 7.9 MB
// Time  Complexity: O(N*N!)
// Space Complexity: O(N*N!)
//https://leetcode-cn.com/problems/permutations-ii/solution/shuang-bai-si-lu-chao-ji-jian-ji-by-tai-yang-tian/
// solution 1,backtrack with swapping elements
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
private:
    vector<vector<int>> res;
    void generatePermutation(vector<int> &nums, int begin) {
        if (begin == nums.size()) {
            res.push_back(nums);
            return;
        }
        generatePermutation(nums, begin + 1);
        for (int i = 0; i < begin; i++) {
            swap(nums[i], nums[begin]);
            generatePermutation(nums, begin + 1);
            swap(nums[i], nums[begin]);
        }
    }
public:
    vector<vector<int>> permute(vector<int> &nums) {
        sort(nums.begin(), nums.end());
        generatePermutation(nums, 0);
        return res;
    }
};